File: pst.rb
#!/usr/bin/ruby -w

require 'etc'
require 'stringio'

def escape(cmd)
  buf = StringIO.new
  cmd.each_char do |c|
    if c == '\\'
      buf << '\\\\'
    elsif c > ' ' and c <= '~'
      buf << c
    else
      buf << '\\%03o' % c.ord
    end
  end
  buf.string
end

def arg_generator(str)
  buf = StringIO.new
  str.each_char do |c|
    if c == "\0"
      s = buf.string.clone
      buf.seek 0
      buf.truncate 0
      yield s
    else
      buf << c
    end
  end
end

def build_proc(pid)
  proc_dir = '/proc/%d' % pid
  stat = File.stat(proc_dir)
  uid = stat.uid
  args = []
  arg_generator(open('%s/cmdline' % proc_dir).read()) { |arg| args << arg }
  proc_stat = open('%s/stat' % proc_dir).read()
  cmd = proc_stat[proc_stat.index('(') + 1 ... proc_stat.rindex(')')]
  ppid = proc_stat[proc_stat.index(')') + 4 .. -1].split()[0].to_i
  return {cmd: cmd, args: args, pid: pid, ppid: ppid, uid: uid}
end

def get_pids(proc_dir='/proc')
  pids = []
  Dir.foreach(proc_dir) do |name|
    pid = name.to_i
    pids << pid if pid != 0
  end
  pids
end

def read_procs(proc_dir='/proc')
  get_pids(proc_dir).collect { |pid| build_proc(pid) }
end

class UsernameCache
  @cache = {}
  def self.get_username(uid)
    begin
      @cache.fetch uid
    rescue KeyError
      name = Etc.getpwuid(uid).name
      @cache[uid] = name
      name
    end
  end
end

def show_proc(procs, depth, last, more_at_depth, last_uid, proc)
  if depth != 0
    (1 .. depth).each do |i|
      more = more_at_depth[i]
      print('  ')
      if i == depth
        print(last ? '`-' : '|-')
      else
        print(more ? '| ' : '  ')
      end
    end
  end
  print("%s,%d%s%s\n" %
        [proc[:cmd], proc[:pid],
         (proc[:uid] != last_uid ?
          ',' + UsernameCache.get_username(proc[:uid]) :
          ''),
         (proc[:args].length > 1 ?
          ' ' + proc[:args][1..-1].collect { |arg| escape arg }.join(' ') :
          '')])
  show_children(procs, depth, more_at_depth, proc[:pid], proc[:uid])
end

def show_children(procs, depth, more_at_depth, parent, last_uid)
  children = procs.select { |proc| proc[:ppid] == parent }
  depth += 1
  items = children.sort_by { |proc| [proc[:cmd], proc[:pid]] }
  if !items.empty?
    last = items[-1]
    rest = items[0...-1]
    rest.each do |item|
      more = more_at_depth.clone
      more[depth] = true
      show_proc(procs, depth, false, more, last_uid, item)
    end
    more = more_at_depth.clone
    more[depth] = false
    show_proc(procs, depth, true, more, last_uid, last)
  end
end

def show_tree(start, procs)
  root = procs.find { |proc| proc[:pid] == start }
  show_proc(procs, 0, true, {}, 0, root)
end

show_tree(1, read_procs())

File: pst.py
#!/usr/bin/env python

import os
import pwd
import sys

from io import StringIO

def escape(cmd):
    buf = StringIO()
    for c in cmd:
        if c == '\\':
            buf.write('\\\\')
        elif c > ' ' and c <= '~':
            buf.write(c)
        else:
            buf.write('\\%03o' % ord(c))
    return buf.getvalue()

def arg_generator(seq):
    buf = StringIO()
    for c in seq:
        if c == '\0':
            s = buf.getvalue()
            buf.seek(0)
            buf.truncate()
            yield s
        else:
            buf.write(c)

def build_proc(pid):
    proc_dir = '/proc/%d' % pid
    stat = os.stat(proc_dir)
    uid = stat.st_uid
    args = list(arg_generator(open('%s/cmdline' % proc_dir).read()))
    proc_stat = open('%s/stat' % proc_dir).read()
    cmd = proc_stat[proc_stat.index('(') + 1 : proc_stat.rindex(')')]
    ppid = int(proc_stat[proc_stat.index(')') + 4 : ].split()[0])
    return {'cmd': cmd, 'args': args, 'pid': pid, 'ppid': ppid, 'uid': uid}

def get_pids(proc_dir='/proc'):
    pids = []
    for name in os.listdir(proc_dir):
        try:
            pids.append(int(name))
        except ValueError:
            pass
    return pids

def read_procs(proc_dir='/proc'):
    return [build_proc(pid) for pid in get_pids(proc_dir)]

def get_username(uid, _cache={}):
    try:
        return _cache[uid]
    except KeyError:
        name = pwd.getpwuid(uid).pw_name
        _cache[uid] = name
        return name

def show_tree(start, procs):
    def show_children(depth, more_at_depth, parent, last_uid):
        children = (proc for proc in procs if proc['ppid'] == parent)
        depth += 1
        items = sorted(children, key=lambda proc: (proc['cmd'], proc['pid']))
        if items:
            last = items[-1]
            rest = items[:-1]
            for item in rest:
                more = more_at_depth.copy()
                more[depth] = True
                show_proc(depth, False, more, last_uid, item)
            more = more_at_depth.copy()
            more[depth] = False
            show_proc(depth, True, more, last_uid, last)
    def show_proc(depth, last, more_at_depth, last_uid, proc):
        if depth != 0:
            for i in range(1, depth + 1):
                more = more_at_depth[i]
                sys.stdout.write('  ')
                if i == depth:
                    sys.stdout.write('`-' if last else '|-')
                else:
                    sys.stdout.write('| ' if more else '  ')
        sys.stdout.write('%s,%d%s%s\n'
                         % (proc['cmd'], proc['pid'],
                            (',' + get_username(proc['uid'])
                             if proc['uid'] != last_uid
                             else ''),
                            (' ' + ' '.join(map(escape, proc['args'][1:]))
                             if len(proc['args']) > 1
                             else '')))
        show_children(depth, more_at_depth, proc['pid'], proc['uid'])
    root = [proc for proc in procs if proc['pid'] == start][0]
    show_proc(0, True, {}, 0, root)

if __name__ == '__main__':
    show_tree(1, read_procs())
